tests/modify.js

```diff
@@ -50,19 +49,39 @@ describe('modify', () => {
         const h1 = hamt.empty
             .modifyHash(0, 'a', () => 3)
             .modifyHash(0, 'b', () => 5);
-    
+
         const h3 = h1.modifyHash(0, 'a', x => x * 2);
         assert.strictEqual(6, h3.getHash(0, 'a'));
         assert.strictEqual(5, h3.getHash(0, 'b'));
-    
+
         const h4 = h3.modifyHash(0, 'b', x => x * 2);
         assert.strictEqual(6, h4.getHash(0, 'a'));
         assert.strictEqual(10, h4.getHash(0, 'b'));
-    
+
         // Non existant
         const h5 = h4.modifyHash(0, 'c', _ => 100);
         assert.strictEqual(6, h5.getHash(0, 'a'));
         assert.strictEqual(10, h5.getHash(0, 'b'));
         assert.strictEqual(100, h5.getHash(0, 'c'));
     });
+
+    it('should provide defaultValue to `f` if entry does not exist on empty map', () => {
+        // TODO: complete
+    });
+
+    it('should provide defaultValue to `f` if entry does not exist on non-empty map', () => {
+
+        const h = hamt.set('a', 3, hamt.empty);
+
+        const NOT_SET = {};
+
+        const transform = (x) => {
+
+            assert.strictEqual(x, NOT_SET);
+
+            return 'b';
+        };
+
+        const h1 = hamt.modify(transform, 'b', NOT_SET, h);
```

Comment:
In the case where `defaultValue` is provided, `f` is actually still called, because, internally, the function is wrapped by:

```js
var defaultValBind = function defaultValBind(f, defaultValue) {
    return function (x) {
        return f(arguments.length == 0 ? defaultValue : x);
    };
};
```

If `key` was not found, then `defaultValue` is passed to `f` for insertion operation (but `f` is executed **like** a modification operation).

--------

In the original implementation, `defaultValue` was always provided, regardless if the user actually passed `defaultValue` or not. The default case where `defaultValue` is not given, then it's value is `void 0`. This subtle difference in semantics would have broken this test: https://github.com/mattbierner/hamt/blob/36e3dffff622d24aba949d7892f2f5d00de65821/tests/modify.js#L29-L34